File: LeetCode/Medium/0260-single-number-iii/0260-single-number-iii.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        if len(nums) ==2:
            return nums

        xor = 0
        for i in nums:
            xor ^= i

        a,b = 0,0
        pos = 0

        pos = 0
        while True:
            if (xor>>pos) & 1:
                break
            pos+=1
        
        for i in nums:
            if (i>>pos) & 1:
                a ^= i
                continue
            b ^= i

        

        return [a,b]
                

File: LeetCode/Medium/0260-single-number-iii/0260-single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int x = 0;
        for(int i=0; i<nums.size(); i++) 
            x ^= nums[i];

        int bitpos = 0;
        int i = 0;
        while(!bitpos)
            bitpos |= x>>i & 1 ? 1<<i : 0,i++;

        int a=0,b=0;

        for(int i=0; i<nums.size(); i++){
            if(nums[i]&bitpos) a ^= nums[i];
            else b ^= nums[i];
        }

        return {a,b};
    }
};